stackslib/src/net/download/nakamoto.rs

```diff
@@ -0,0 +1,4283 @@
+                .expect("FATAL: burnchain tip is before system start");
+
+            let mut wanted_tenures = vec![];
+            Self::update_wanted_tenures_for_reward_cycle(
+                sort_rc,
+                sort_tip,
+                sortdb,
+                &mut wanted_tenures,
+            )?;
+
+            test_debug!(
+                "initial wanted_tenures (rc {}): {:?}",
+                sort_rc,
+                &wanted_tenures
+            );
+            self.wanted_tenures = wanted_tenures;
+            self.reward_cycle = sort_rc;
+        }
+        Ok(())
+    }
+
+    /// Determine if the set of `TenureStartEnd`s represents available but unfetched data.  Used to
+    /// determine whether or not to update the set of wanted tenures -- we don't want to skip
+    /// fetching wanted tenures if they're still available!
+    pub(crate) fn have_unprocessed_tenures<'a>(
+        first_nakamoto_rc: u64,
+        completed_tenures: &HashSet<ConsensusHash>,
+        prev_wanted_tenures: &[WantedTenure],
+        tenure_block_ids: &HashMap<NeighborAddress, AvailableTenures>,
+        pox_constants: &PoxConstants,
+        first_burn_height: u64,
+        inventory_iter: impl Iterator<Item = &'a NakamotoTenureInv>,
+    ) -> bool {
+        if prev_wanted_tenures.len() == 0 {
+            return true;
+        }
+
+        // the anchor block for prev_wanted_tenures must not only be processed, but also we have to
+        // have seen an inventory message from the subsequent reward cycle.  If we can see
+        // inventory messages for the reward cycle after `prev_wanted_rc`, then the former will be
+        // true
+        let prev_wanted_rc = prev_wanted_tenures
+            .first()
+            .map(|wt| {
+                pox_constants
+                    .block_height_to_reward_cycle(first_burn_height, wt.burn_height)
+                    .expect("FATAL: wanted tenure before system start")
+            })
+            .unwrap_or(u64::MAX);
+
+        let cur_wanted_rc = prev_wanted_rc.saturating_add(1);
+
+        let mut has_prev_inv = false;
+        let mut has_cur_inv = false;
+        for inv in inventory_iter {
+            if prev_wanted_rc < first_nakamoto_rc {
+                // assume the epoch 2.x inventory has this
+                has_prev_inv = true;
+            } else if inv.tenures_inv.get(&prev_wanted_rc).is_some() {
+                has_prev_inv = true;
+            }
+
+            if cur_wanted_rc < first_nakamoto_rc {
+                // assume the epoch 2.x inventory has this
+                has_cur_inv = true;
+            } else if inv.tenures_inv.get(&cur_wanted_rc).is_some() {
+                has_cur_inv = true;
+            }
+        }
+
+        if !has_prev_inv || !has_cur_inv {
+            test_debug!("No peer has an inventory for either the previous ({},{}) or current ({},{}) wanted tenures", prev_wanted_rc, has_prev_inv, cur_wanted_rc, has_cur_inv);
```

Comment:
```suggestion
            warn!("No peer has an inventory for either the previous ({}: available = {}) or current ({}: available = {}) reward cycles for which we want tenures", prev_wanted_rc, has_prev_inv, cur_wanted_rc, has_cur_inv);
```

Reply:
This can get chatty if the peer currently has no neighbors (e.g. because its network connection got severed).  Are you sure?

Reply:
Ah, no -- in that case, is it easy to check if there's no peers? I would want to see warnings if inventory sync isn't able to make progress because of this, but I wouldn't want this to spam things like a single-node regtest env

Reply:
You'd see errors in the inventory state machine, not here.